src/librustc_typeck/astconv.rs

```diff
@@ -1364,15 +1392,61 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         param_ty: Ty<'tcx>,
         ast_bounds: &[hir::GenericBound<'_>],
         bounds: &mut Bounds<'tcx>,
+        skip_self_param_evaluation: bool,
     ) {
         let mut trait_bounds = Vec::new();
         let mut region_bounds = Vec::new();
 
+        /// Walk each bound to see if they reference an associated type of `Self` without
+        /// specifying a trait (using a Resolved path `Self::Foo` instead of fully qualified path
+        /// `<Self as Bar>::Foo`) and if so, skip them if `skip_self_param_evaluation` is `true`.
+        /// This way, we avoid cycle errors and are able to resolve the associated type
+        /// successfully.
+        struct SelfFinder(bool);
+
+        impl<'v> Visitor<'v> for SelfFinder {
+            type Map = intravisit::ErasedMap<'v>;
+
+            fn nested_visit_map(&mut self) -> intravisit::NestedVisitorMap<Self::Map> {
+                intravisit::NestedVisitorMap::None
+            }
+
+            fn visit_ty(&mut self, t: &'v hir::Ty<'v>) {
+                if let hir::TyKind::Path(hir::QPath::TypeRelative(
+                    hir::Ty {
+                        kind:
+                            hir::TyKind::Path(
+                                hir::QPath::Resolved(
+                                    None,
+                                    hir::Path { res: Res::SelfTy(Some(_), None), .. },
+                                ),
+                                ..,
+                            ),
+                        ..
+                    },
+                    _,
+                )) = t.kind
+                {
+                    self.0 = true;
+                }
+                intravisit::walk_ty(self, t)
+            }
+        }
+
         let constness = self.default_constness_for_trait_bounds();
         for ast_bound in ast_bounds {
             match *ast_bound {
-                hir::GenericBound::Trait(ref b, hir::TraitBoundModifier::None) => {
-                    trait_bounds.push((b, constness))
+                hir::GenericBound::Trait(ref b, modif @ hir::TraitBoundModifier::None) => {
+                    let mut visitor = SelfFinder(false);
+                    visitor.visit_poly_trait_ref(b, modif);
+
+                    let is_self_param = match param_ty.kind {
+                        ty::Param(param) if param.name == kw::SelfUpper => true,
+                        _ => false,
+                    };
+                    if !(visitor.0 && skip_self_param_evaluation && is_self_param) {
```

Comment:
Ideally instead of just skipping super traits we would synthesize a new PolyTraitRef with a TraitRef with paramus set to some placeholder to avoid the weird edge cases I mentioned before, but couldn't see a good way to do that to a hir node.